### PR TITLE
C++2b / C++23 / C17 - Clang macOS/iOS - MSVC VS2022 - VS

### PR DESCRIPTION
**Pull request by danoli3**
## C++ 23 / C17
- Clang macOS/iOS 
- MSVC VS2022 - VS 

This PR Updating Templates. 

Apothecary already building bleeding libraries with C++23 standard / C++2b https://github.com/openframeworks/apothecary/pull/390

### CODE REVIEW

**Comment by danoli3** [comment]
Not sure why C++23 is not accepted in VS2022 with latest. It might be written with c++2b instead of c++23 

https://learn.microsoft.com/en-us/cpp/overview/what-s-new-for-visual-cpp-in-visual-studio?view=msvc-170
As its meant to be included in 2023 .. Screenshot 2024-07-17 at 8 44 27 PM



https://learn.microsoft.com/en-us/cpp/overview/install-c17-support?view=msvc-170&source=recommendations

**Comment by danoli3** [comment]
```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error	C2280	'std::basic_ostream<char,std::char_traits<char>> &std::operator <<<std::char_traits<char>>(std::basic_ostream<char,std::char_traits<char>> &,const wchar_t *)': attempting to reference a deleted function	openframeworksLib	\\mac\Home\SOURCE\openFrameworksmacOS\libs\openFrameworks\utils\ofLog.h	509		
```

fixing now with:
on latest with c++latest used (23) 
also set c17

**Comment by danoli3** [comment]
MSVC some issues in ofLog() 
``
template <class T>
ofLog & operator<<(const T & value) {
	message << value << getPadding();
	return *this;
}
```

```
Severity	Code	Description	Project	File	Line	Suppression State	Details
Error	MSB4025	The project file could not be loaded. Could not find file 'C:\srx\openFrameworks\examples\templates\emptyExample\emptyExample.vcxproj'.	emptyExample	C:\srx\openFrameworks\examples\templates\emptyExample\emptyExample.vcxproj	1		
Error	C2280	'std::basic_ostream<char,std::char_traits<char>> &std::operator <<<std::char_traits<char>>(std::basic_ostream<char,std::char_traits<char>> &,const wchar_t *)': attempting to reference a deleted function	openframeworksLib	C:\srx\openFrameworks\libs\openFrameworks\utils\ofLog.h	509		
Error	C2088	built-in operator '<<' cannot be applied to an operand of type 'std::stringstream'	openframeworksLib	C:\srx\openFrameworks\libs\openFrameworks\utils\ofLog.h	509		
```

<img width="927" alt="Screenshot 2024-07-18 at 1 48 00 AM" src="https://github.com/user-attachments/assets/78cac877-1bde-46ab-86cc-3ce7e6cd8ac6">

<img width="1067" alt="Screenshot 2024-07-18 at 1 47 13 AM" src="https://github.com/user-attachments/assets/e7b07218-2df2-49a5-8b6e-2973b0f5ae69">

**Comment by 2bbb** [comment]
https://learn.microsoft.com/en-us/cpp/overview/cpp-conformance-improvements-2019?view=msvc-170

> Traditionally, inserting a wchar_t into a std::ostream, and inserting char16_t or char32_t into a std::ostream or std::wostream, outputs its integral value. Inserting pointers to those character types outputs the pointer value. Programmers don't find either case intuitive. They often expect the standard library to transcode the character or null-terminated character string instead, and to output the result.
> The C++20 proposal P1423R3 adds deleted stream insertion operator overloads for these combinations of stream and character or character pointer types. Under /std:c++20 or /std:c++latest, the overloads make these insertions ill-formed, instead of behaving in what is likely an unintended manner. The compiler raises error C2280 when one is found. You can define the "escape hatch" macro _HAS_STREAM_INSERTION_OPERATORS_DELETED_IN_CXX20 to 1 to restore the old behavior. (The proposal also deletes stream insertion operators for char8_t. Our standard library implemented similar overloads when we added char8_t support, so the "wrong" behavior has never been available for char8_t.)

**Comment by danoli3** [comment]
_HAS_STREAM_INSERTION_OPERATORS_DELETED_IN_CXX20 macro fixes VS2022 c++latest for the moment with ofLog issue for now. Might be a good idea to get this in as well

**Comment by danoli3** [comment]
- Fixed Bash ci bleeding project generator commands for Windows.
- Disabled Linux64 -2204 (now just 2404)
# Updated VS2022 project files (template for PG / Core OF Project) 
- Updated  support for Unicode (C/C++), including paths for vs project linking 
- Updated for C++23 / C++ Modules 
- Updated for C17 (latest C standard) - fixes C issues

File: addons/ofxKinect/libs/libfreenect/src/loader.c [hunk 1]
```diff
@@ -169,7 +169,7 @@ FN_INTERNAL int upload_firmware(fnusb_dev* dev, char * filename) {
 				if (!home)
 					continue;
 				int homelen = strlen(home);
-				char* dotfolder = "/.libfreenect";
+				char dotfolder[14] = "/.libfreenect";
```
Comment (danoli3) [inline addons/ofxKinect/libs/libfreenect/src/loader.c:172]: fixes C17 issues  with this addon - as char* became const in new standard. this fixes it

File: .github/workflows/of.yml [hunk 2]
```diff
@@ -118,15 +118,16 @@ jobs:
...
-      # - name: Build DEBUG allAddonsExample
-      #   working-directory: ${{env.GITHUB_WORKSPACE}}
-      #   run:
-      #     msbuild examples/templates/allAddonsExample/allAddonsExample.vcxproj /p:configuration=debug /p:platform=${{ matrix.platform }} /p:PlatformToolset=v143
-
+          msbuild examples/templates/emptyExample/emptyExample.vcxproj /p:configuration=release /p:platform=${{ matrix.platform }} /p:PlatformToolset=v143
+      - name: Build DEBUG and RELEASE allAddonsExample
+        working-directory: ${{env.GITHUB_WORKSPACE}}
+        run: |
+          msbuild examples/templates/allAddonsExample/allAddonsExample.vcxproj /p:configuration=debug /p:platform=${{ matrix.platform }} /p:PlatformToolset=v143
```
Comment (danoli3) [inline .github/workflows/of.yml:129]: added release and addon tests

File: addons/ofxKinect/libs/libfreenect/src/usb_libusb10.c [hunk 3]
```diff
@@ -585,7 +585,8 @@ FN_INTERNAL int fnusb_open_subdevices(freenect_device *dev, int index)
 			}
 			else
 			{
-				res = upload_firmware(&dev->usb_audio, "audios.bin");
+				char audiosBin[11] = "audios.bin";
```
Comment (danoli3) [inline addons/ofxKinect/libs/libfreenect/src/usb_libusb10.c:588]: fixes C17 issues

File: libs/openFrameworks/app/ofAppEGLWindow.cpp [hunk 4]
```diff
@@ -409,10 +409,10 @@ void ofAppEGLWindow::setup(const ofAppEGLWindowSettings & _settings) {
 
 	// X11 check
 	// char * pDisplay;
-	// pDisplay = getenv ("DISPLAY");
+	// pDisplay = ofGetEnv("DISPLAY");
 	// bool bIsX11Available = (pDisplay != NULL);
 
-	bool bIsX11Available = getenv("DISPLAY") != NULL;
+	bool bIsX11Available = ofGetEnv("DISPLAY") != NULL;
```
Comment (danoli3) [inline libs/openFrameworks/app/ofAppEGLWindow.cpp:415]: use oF wrapper with exception handler

File: libs/openFrameworks/utils/ofUtils.cpp [hunk 5]
```diff
@@ -1132,7 +1132,7 @@ std::string ofGetEnv(const std::string & var, const std::string defaultValue) {
 		return defaultValue;
 	}
 #else
-	auto value = getenv(var.c_str());
+	auto value = std::getenv(var.c_str());
```
Comment (danoli3) [inline libs/openFrameworks/utils/ofUtils.cpp:1135]: explicit

File: scripts/ci/vs/install_bleeding.sh [hunk 6]
```diff
@@ -15,25 +14,39 @@ cd projectGenerator
...
+#unzip projectGenerator-vs-gui.zip 2> /dev/null
+#rm projectGenerator-vs-gui.zip
+
+cd "${OF_ROOT}"
 
-# downloader https://github.com/openframeworks/projectGenerator/releases/download/nightly/projectGenerator-vs-gui.zip 2> /dev/null
-# unzip projectGenerator-vs-gui.zip 2> /dev/null
-# rm projectGenerator-vs-gui.zip
+OFW_ROOT=$(cygpath -w "$PWD")
+PG_OFX_PATH="${OF_ROOT}\\projectGenerator\\projectGenerator.exe"
```
Comment (danoli3) [inline scripts/ci/vs/install_bleeding.sh:27]: this fixed the Ci PG for MSBUILD paths

File: scripts/ci/vs/install_bleeding.sh [hunk 7]
```diff
@@ -15,25 +14,39 @@ cd projectGenerator
...
+    ""
+    "ofxAssimpModelLoader,ofxGui,ofxKinect,ofxNetwork,ofxOpenCv,ofxOsc,ofxSvg,ofxThreadedImageLoader,ofxXmlSettings"
+)
+
+echo "Updating projects with PG at:${PG_OF_PATH}"
+
+for i in "${!PROJECTS[@]}"; do
+    PROJECT=${PROJECTS[i]}
+    ADDON=${ADDONS[i]}
+    OPTIONS="-o\"${OF_ROOT}\" -v -a\"$ADDON\" -p\"vs\"  -t\"\" \"${OF_ROOT}\\${PROJECT}\""
```
Comment (danoli3) [inline scripts/ci/vs/install_bleeding.sh:47]: previously was missing -p and addons

**Review by danoli3 — COMMENTED** [review]
Building amazing on VS2022

**Comment by danoli3** [comment]
C++23 lets go

**Comment by danoli3** [comment]
This solves issues building / fixes projectGenerator download scripts. Fixes 2204 actions.

**Pull request merged**